File: 0001-0500/0084_largestRectangleArea.py
class Solution:
    def largestRectangleArea(self, heights: List[int]) -> int:
        maxArea = 0
        heights.append(0)
        stack = [-1]
        for i in range(len(heights)):
            while heights[i] < heights[stack[-1]]:
                height = heights[stack.pop()]
                width = i - stack[-1] - 1
                maxArea = max(maxArea, height * width)
            stack.append(i)
        return maxArea

# Alternative solution

class Solution:
    def largestRectangleArea(self, heights: List[int]) -> int:
        stack = []
        max_area = 0
        i = 0

        while i < len(heights):
            max_area = max(max_area, heights[i])
            width = 1

            while stack and stack[-1][0] >= heights[i]:
                h, w = stack.pop()
                if stack and stack[-1][0] >= heights[i]:
                    stack[-1] = [min(stack[-1][0], h), stack[-1][1] + w]
                    max_area = max(max_area, stack[-1][0] * stack[-1][1])
                else:
                    width += w
                    max_area = max(max_area, min(h, heights[i]) * width)
                    break
        
            stack.append([heights[i], width])
            i += 1

        while len(stack) > 1:
            h, w = stack.pop()
            stack[-1] = [stack[-1][0], stack[-1][1] + w]
            max_area = max(max_area, stack[-1][0] * stack[-1][1])
            
        return max_area

# Alternative solution

class Solution:
    def largestRectangleArea(self, heights: List[int]) -> int:
        stack = [[-float('inf'), 0]]
        max_area = 0

        for height in heights:
            width = 0
            h = stack[-1][0]

            while stack[-1][0] > height:
                h, w = stack.pop()
                width += w
                max_area = max(max_area, h * width)

            max_area = max(max_area, height * (width + 1))
            stack.append([height, width + 1])
        
        width = 0

        while len(stack) > 1:
            h, w = stack.pop()
            width += w
            max_area = max(max_area, h * width)
        
        return max_area

File: 0001-0500/0084_largestRectangleArea.java
class Solution {
    public int largestRectangleArea(int[] heights) {
        Stack<List<Integer>> stack = new Stack<>();
        int n = heights.length;
        int maxArea = 0;

        for (int i = 0; i < n; i++) {
            int height = heights[i];
            int width = 1;

            while (!stack.empty() && stack.peek().get(0) >= height) {
                List<Integer> x = stack.pop();

                if (!stack.empty() && stack.peek().get(0) >= height) {
                    List<Integer> y = stack.pop();
                    stack.push(List.of(y.get(0), y.get(1) + x.get(1)));
                    maxArea = Math.max(maxArea, stack.peek().get(0) * stack.peek().get(1));
                } else {
                    width += x.get(1);
                    maxArea = Math.max(maxArea, height * width);
                }
            }

            stack.push(List.of(height, width));
            maxArea = Math.max(maxArea, height * width);
        }

        int w = stack.pop().get(1);

        while (!stack.empty()) {
            List<Integer> z = stack.pop();
            w += z.get(1);
            maxArea = Math.max(maxArea, z.get(0) * w);
        }

        return maxArea;
    }
}

// Alternative solution

class Solution {
    public int largestRectangleArea(int[] heights) {
        Stack<List<Integer>> stack = new Stack<>();
        int n = heights.length;
        int maxArea = 0;

        for (int i = 0; i < n; i++) {
            int height = heights[i];
            int width = 0;

            while (!stack.empty() && stack.peek().get(0) >= height) {
                List<Integer> popped = stack.pop();
                width += popped.get(1);
                maxArea = Math.max(maxArea, popped.get(0) * width);
            }

            stack.push(List.of(height, width + 1));
            maxArea = Math.max(maxArea, height * (width + 1));
        }

        int w = stack.pop().get(1);

        while (!stack.empty()) {
            List<Integer> z = stack.pop();
            w += z.get(1);
            maxArea = Math.max(maxArea, z.get(0) * w);
        }

        return maxArea;
    }
}
